gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/filter/RememberDeviceFilter.java

```diff
@@ -38,18 +38,18 @@ public Boolean get() {
         // If Adaptive MFA is active and KO (rule == false) we bypass this filter
         // Because the device could be known and skip MFA
         final boolean mfaSkipped = context.isMfaSkipped();
-        var rememberDeviceSettings = context.getRememberDeviceSettings();
 
-        final boolean userStronglyAuth = context.isUserStronglyAuth();
-        if (!mfaSkipped && context.isAmfaActive()) {
-            if (!context.isAmfaRuleTrue() && !rememberDeviceSettings.isSkipRememberDevice()) {
-                return UNSAFE;
-            } else if (context.isAmfaRuleTrue() && !userStronglyAuth) {
+        // We might want to evaluate Adaptive MFA
+        var rememberDeviceSettings = context.getRememberDeviceSettings();
+        if (!mfaSkipped && context.isAmfaActive() && !context.isAmfaRuleTrue()) {
+            if ((context.isUserStronglyAuth() && !context.isStepUpActive()) || (rememberDeviceSettings.isSkipRememberDevice() && context.deviceAlreadyExists())) {
                 return SAFE;
```

Comment:
it is expected behavior only for REQUIRED mfa

Reply:
only for adaptive mfa yes